arch/src/aarch64/fdt.rs

```diff
@@ -547,107 +542,138 @@ fn create_devices_node<T: DeviceInfoForFdt + Clone + Debug, S: ::std::hash::Buil
     // See Documentation/devicetree/bindings/pci/host-generic-pci.txt in the kernel
     // and https://elinux.org/Device_Tree_Usage.
-
-    // EDK2 requires the PCIe high space above 4G address.
-    // The actual space in CLH follows the RAM. If the RAM space is small, the PCIe high space
-    // could fall bellow 4G.
-    // Here we put it above 512G in FDT to workaround the EDK2 check.
-    // But the address written in ACPI is not impacted.
-    let pci_device_base_64bit: u64 = if cfg!(feature = "acpi") {
-        pci_device_base + PCI_HIGH_BASE
-    } else {
-        pci_device_base
-    };
-    let pci_device_size_64bit: u64 = if cfg!(feature = "acpi") {
-        pci_device_size - PCI_HIGH_BASE
-    } else {
-        pci_device_size
-    };
-
-    let ranges = [
-        // io addresses
-        0x1000000,
-        0_u32,
-        0_u32,
-        (MEM_PCI_IO_START.0 >> 32) as u32,
-        MEM_PCI_IO_START.0 as u32,
-        (MEM_PCI_IO_SIZE >> 32) as u32,
-        MEM_PCI_IO_SIZE as u32,
-        // mmio addresses
-        0x2000000,                                // (ss = 10: 32-bit memory space)
-        (MEM_32BIT_DEVICES_START.0 >> 32) as u32, // PCI address
-        MEM_32BIT_DEVICES_START.0 as u32,
-        (MEM_32BIT_DEVICES_START.0 >> 32) as u32, // CPU address
-        MEM_32BIT_DEVICES_START.0 as u32,
-        (MEM_32BIT_DEVICES_SIZE >> 32) as u32, // size
-        MEM_32BIT_DEVICES_SIZE as u32,
-        // device addresses
-        0x3000000,                            // (ss = 11: 64-bit memory space)
-        (pci_device_base_64bit >> 32) as u32, // PCI address
-        pci_device_base_64bit as u32,
-        (pci_device_base_64bit >> 32) as u32, // CPU address
-        pci_device_base_64bit as u32,
-        (pci_device_size_64bit >> 32) as u32, // size
-        pci_device_size_64bit as u32,
-    ];
-    let bus_range = [0, 0]; // Only bus 0
-    let reg = [PCI_MMCONFIG_START.0, PCI_MMCONFIG_SIZE];
-
-    let pci_node = fdt.begin_node("pci")?;
-    fdt.property_string("compatible", "pci-host-ecam-generic")?;
-    fdt.property_string("device_type", "pci")?;
-    fdt.property_array_u32("ranges", &ranges)?;
-    fdt.property_array_u32("bus-range", &bus_range)?;
-    fdt.property_u32("#address-cells", 3)?;
-    fdt.property_u32("#size-cells", 2)?;
-    fdt.property_array_u64("reg", &reg)?;
-    fdt.property_u32("#interrupt-cells", 1)?;
-    fdt.property_null("interrupt-map")?;
-    fdt.property_null("interrupt-map-mask")?;
-    fdt.property_null("dma-coherent")?;
-    fdt.property_u32("msi-parent", MSI_PHANDLE)?;
-
-    if let Some(virtio_iommu_bdf) = virtio_iommu_bdf {
-        // See kernel document Documentation/devicetree/bindings/pci/pci-iommu.txt
-        // for 'iommu-map' attribute setting.
-        let iommu_map = [
+    // In multiple PCI segments setup, each PCI segment needs a PCI node.
+    for pci_device_info_elem in pci_device_info.iter() {
+        // EDK2 requires the PCIe high space above 4G address.
```

Comment:
The input data is already aligned to 4GiB, we can remove this adjustment.

See: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/device_manager.rs#L946

Reply:
As per discussion, the proper fix should be https://github.com/cloud-hypervisor/cloud-hypervisor/pull/3391/commits/e5785133ece1804c8480bce2d6b9ffb0921aef39

Reply:
Done